### ISSUE TITLE
Really slow in Chrome on Windows

### ISSUE THREAD

**Issue by ksmithut** [highlightjs/highlight.js#2326]
I have a fairly wittled down example of some slowness I'm seeing.

https://gist.github.com/ksmithut/31b66e0bc20e4bc0ed504f2dea4bfa2c

The content is rather large (around 50k characters of html), but in Chrome on Windows, it's unbearably slow to select any text, and even scrolling takes a long time. I tried this on Linux on Firefox and ~Chromium with no issue~ Chrome on Linux was also slow, and I tried Firefox on Windows and it was also just fine. Chrome on Windows is the one having the issue.

I'm fine if the answer is to not highlight anything that long, but a good performant recommendation would be very appreciated.

**Comment by yyyc514** [comment]
Does this really have anything to do with Highlight.js or is it merely the rendered content? (though I suppose you could still argue we generated poor content)...

If you just save the same webpage statically without us rendering the content and then load it does it have the same slowness?

**Comment by yyyc514** [comment]
Seems pretty normal on latest Chrome on Mac here.

**Comment by yyyc514** [comment]
> I'm fine if the answer is to not highlight anything that long, but a good performant recommendation would be very appreciated.

I think this (in general) has more to do with the browser than with anything else.  A old computer barely running a modern browser is going to have a problem with a HUGE webpage with tons of different tags and classes... but if some browsers work just fine and only one is slow, then that's something else entirely.

If someone else can provide more details, that'd be great.  I don't use Windows.

**Comment by ksmithut** [comment]
For added context, these were both computers with 64gb of RAM and a less than 2 year old processor (i9 for both) also with decent graphics cards (RTX 2060 and GTX 1080ti). I don't doubt that hardware acceleration wasn't enabled for those pages, but I was able to work up an example with just under a kilobyte of data using the json syntax, and it was also having slowdown issues. I don't doubt that it's something to do with Chrome, especially given that I gave it pretty modern specs, and that it works great on any other browser. If it were any other browser I probably wouldn't have even filed it, but since Chrome is one of the most popular browser among our users, it brought up some concern. I'll investigate some alternatives and see what crops up.

**Comment by yyyc514** [comment]
Ok. Probably Chrome then. If you find something specific or a trigger it’d be helpful to know but ultimately right now this sounds like some sort of Chrome bug.

**Comment by yyyc514** [comment]
@ksmithut Did you try it with different CSS themes or try hacking the CSS to see if it's some sort of weird rendering issue?

**Comment by tiyunchen** [comment]
my problem is window performance is better than mac, window fps is 60 while in mac just only 24~29,so sad

**Comment by yyyc514** [comment]
@tiyunchen What Mac (hardware) and what OS version?

The early Retina hardware was pretty bad (performance wise) but I haven't seen any real issues recently (though I don't use Chrome everyday).  I'm on hardware from 2015 (Retina MBP 15") and Mojave. 

Unfortunately nothing we can really do unless someones shows us this is actually a Highlight.js specific bug vs just an issue with general browser rendering performance.

**Comment by yyyc514** [comment]
> but I was able to work up an example with just under a kilobyte of data using the json syntax, and it was also having slowdown issues.

I you can produce a standalone self-enclosed test (static HTML file) that others could look at to confirm the issue or play around with that might be helpful.

**Comment by ksmithut** [comment]
The example above is a pretty simple one that illustrates the problem. To open up in a browser, run:

**Comment by ksmithut** [comment]
I've moved things to a single file, should be fairly easy to reproduce:

```sh
git clone https://gist.github.com/ksmithut/31b66e0bc20e4bc0ed504f2dea4bfa2c slowness
cd slowness
open index.html # May need to copy the url and paste it into Chrome if your default browser is not chrome
```

There's a theme selector at the top to try out different themes, but from the dozen or so that I tried, the problem remains in chrome. I tried it out on my mac (15" 2018, 32gb RAM, 2.9 GHz 6-core). Just want to clarify, that the problem really only occurs when you start trying to select text. Perhaps there is some event listener being attached to the various nodes or something is triggering on mouse move that is blocking? I'm also not convinced that Chrome is blameless here, but I'm just seeking to understand what's going on in Chrome that is causing this slowness.

**Comment by yyyc514** [comment]
> Just want to clarify, that the problem really only occurs when you start trying to select text.

Should have bolded and underlined that the first time, or else it's just my fault for missing it. :-)  Ok, now I see the behavior at least. :)  Let me play around with.

**Comment by yyyc514** [comment]
@ksmithut Thanks for bringing this to our attention.  Most certainly looks to be a Chrome bug to me and a few other people are reporting similar issues:

- https://support.google.com/chrome/thread/23344225?hl=en
- https://meta.stackoverflow.com/questions/392404/copy-and-paste-scrolling-freezes-browser-on-a-particular-question

I did play around with your sample but the only thing I found to speed it up was to mostly disable the highlighting... seems Chrome has some bug where selecting when there are lots of HTML nodes is super slow.  With just one large text node it worked fine.

The sample also works silky smooth in Safari for me.

Closing this as seems it's unfortunately nothing we can solve.
